<launch>
    <arg name="run_mloam" default="true" />
    <arg name="run_mloam_odom" default="true" />
    <arg name="run_mloam_mapping" default="true" />    
    <arg name="run_aloam" default="false" />
    <arg name="run_floam" default="false" />
    <arg name="run_legoloam" default="false" />

    <arg name="data_source" default="bag"/>
    <arg name="result_save" default="false"/>
    <arg name="output_path" default="/home/jjiao/trajectory_results/simu_jackal_mloam/xxxxx"/>
    <arg name="with_ua" default="true" /> <!-- awareness of uncertainty propagation-->    
    <arg name="gf_method" default="wo_gf" /> <!-- gd_fix, gd_float, rnd, fps, wo_gf-->
    <arg name="gf_ratio_ini" default="1" />
    <arg name="gnc" default="false" />
    <arg name="debug_mode" default="false" />
    <arg name="loss_mode" default="huber" /> <!-- huber, gmc-->

    <arg name="config_file" default="$(find mloam)/config/config_simu_jackal.yaml" />
    <group if="$(arg run_mloam)">
        <group if="$(arg run_mloam_odom)">
            <node pkg="mloam" type="mloam_node_sr" name="mloam_node_sr" 
                args="-config_file=$(arg config_file)
                      -result_save=$(arg result_save)
                      -output_path=$(arg output_path)" output="screen">
            </node>
        </group>
        <group if="$(arg run_mloam_mapping)">
            <node pkg="mloam" type="lidar_mapper_keyframe" name="lidar_mapper_keyframe"
                args="-config_file=$(arg config_file)
                      -result_save=$(arg result_save) 
                      -output_path=$(arg output_path)
                      -with_ua=$(arg with_ua)
                      -gf_method=$(arg gf_method)
                      -gf_ratio_ini=$(arg gf_ratio_ini)" output="screen">
                <remap from="/laser_odom" to="/laser_odom_0"/>
            </node>
        </group>
    </group>
    
    <!-- publish static tf -->
    <!-- x y z yaw pitch roll -->
    <!-- x y z qx qy qz qw -->
    <!-- world_base: the init position of robot; world_base->world, world->laser_0, aft_mapped->base_link -->
    <node name="static_tf_world_base_into_world_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0.266 0.734 -0.216 0 0 0.976 world_base world 10"/>
    <node name="static_tf_world_sensor_into_base_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 world laser_0 10"/>
    <node name="static_tf_laser_into_laser_left" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 laser_0 velo_left 10"/>        
    <node name="static_tf_laser_into_laser_right" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 laser_1 velo_right 10"/>               
    <node name="static_tf_base_sensor_into_base_link" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 aft_mapped base_link 10"/>       

    <!-- merge cloud -->
    <group unless="$(arg run_mloam)">
        <node pkg="mloam_test" type="test_merge_pointcloud_sr" 
            name="test_merge_pointcloud_sr" args="$(arg config_file)" output="screen" />      
        <node name="static_tf_world_sensor_into_camera" pkg="tf" type="static_transform_publisher"
            args="0 0 0 0 0 0 1 world camera_init 10"/>            
    </group>

    <!-- run aloam  -->
    <group if="$(arg run_aloam)">
        <include file="$(find aloam_velodyne)/launch/aloam_exp.launch">
            <arg name="output_path" value="$(arg output_path)"/>
            <arg name="skip_frame" value="2" />
            <arg name="mapping_line_res" value="0.2" />
            <arg name="mapping_plane_res" value="0.4" />
            <arg name="s_line" value="16" />
        </include>       
   </group>        

    <!-- run floam  -->
    <group if="$(arg run_floam)">
        <include file="$(find floam)/launch/floam_exp.launch">
            <arg name="output_path" value="$(arg output_path)"/>
            <arg name="s_line" value="16" />
            <arg name="map_res" value="0.4" />
        </include>       
   </group>        

    <!-- run legoloam  -->
    <group if="$(arg run_legoloam)">
        <include file="$(find lego_loam)/launch/legoloam_exp.launch">
            <arg name="output_path" value="$(arg output_path)"/>
        </include>
    </group>          

    <!-- run bag -->
    <arg name="bag_file_use" default="true" />
    <arg name="bag_file" default="/Monster/dataset/lidar_calibration/mloam_dataset/SR01.bag" />

    <param name="use_sim_time" value="true" />
    <group if="$(arg bag_file_use)">
        <node pkg="rosbag" type="play" name="rosbag" 
            args="--clock -r 2 $(arg bag_file)" required="true"/>    
    </group>    

    <!-- <include file="$(find rviz_car_model)/launch/default.launch" >
        <arg name="fixed_frame" value="base_link" />
        <arg name="transform" value="0 0.266 0.734 -0.216 0 0 0.976" />
    </include>     -->

</launch>
